### PR TITLE
TF Estimator Trainer Template for CloudML

### PR DESCRIPTION
**Pull request by ksalama**
This is the initial pull request to add my **TensorFlow Trainer Template** to the cloudml-samples repo.

This pull request includes the **template**, and initial **Readme.md** file that describes the purpose of the template and the various modules included in the template. 

Subsequent pull request will include **examples** of adapting this template to different datasets. Also, the Readme.md will include more **detailed description**/walkthroughs of how to use the template.

### CODE REVIEW

File: cloudml-template/README.md [hunk 1]
```diff
@@ -0,0 +1,24 @@
+# TensorFlow Trainer Package Template
```
Comment (nkashy1) [inline cloudml-template/README.md]: @ksalama : It seems as if this could become a good template for users who want to deploy jobs involving the canned estimators to ML Engine. There are a couple of things that I think are missing for this purpose:

1. Guidance on submitting modified versions of this template (using `config.yaml`) to ML Engine

1. An example which shows users how to turn this template into a working job. There is a lot of code here, and although the table in the README is very helpful, someone using the template will still have to spend some time reading through the code, as well.

Although I have not tried using it for this purpose, I have a feeling that the template does not adapt well to the task of setting up training jobs with custom estimators or even with hand-written training loops with input data potentially requiring some preprocessing, being zipped from multiple sources, etc. In general, keeping this template up-to-date with the changes tensorflow is making to its input pipelines seems like it would be a lot of work and, if we are pitching it as a general template for the submission of training jobs to ML Engine, this is work we would have to do.

For that reason, I suggest not presenting this as a generic trainer template, but rather as a template for canned estimator usage or for common data science problems.

**Comment by ksalama** [comment]
@nkashy1 

Hello Neeraj

Thank you so much for your valuable input. Let first describe what I have added in this commit:

- Examples of how you can adapt the template to your dataset, including:
-- Census Income Prediction - an example of classification with canned estimator.
-- Housing Price Estimation - an example of regression with canned estimator.
-- Baby Weight Estimation - an example of regression with custom estimator (model function).
- Scripts to show how to:
-- Train the model locally.
-- Submit a training job to CMLE.
-- Deploy a trained model to CMLE.
-- Example for config.yaml file (for hyper-parameter tuning + custom scaling tier).

Now, let me address your comments, and recommend next steps

**Script to Submit the a Concrete Implementation of Template** - As mentioned, I added these scripts, along with a config.yaml file template.

**Code Examples** - As mentioned, I have added examples (along with their related datasets) of how the template can be adapted for a given dataset and a particular ML task.

**Walk-through Example in the README*** - Yes I agree. I need to add a step by step example to describe how, for example, I adapted the template for the Census dataset to perform classification.

**Presenting the Template as a Generic TF Training Template*** - I totally agree with you that this template is not suitable for hand-written training loops. However, I think it should be suitable for any **estimator**-based model, including both **canned** and **custom**  estimators (i.e., using a custom model function). An example of a custom estimator is the Baby Weight Estimation example added to last commit. I can implement more examples to demonstrate this (please see these [tutorials](https://github.com/GoogleCloudPlatform/tf-estimator-tutorials) that I created, specially the image & text analysis ones, which uses the sample logic/flow of template, but in notebooks. I can create additional example, using the template, for Image & Text Classification).

In addition, the template should also be suitable for custom preprocessing during the reading data. the **input.py** module has a template function - process_features() - that is applied during reading the data in the _input_fn(). The user can implement the process_features() function to apply custom feature processing (this is also described in the comment of this function in the template . Example of custom feature processing are shown in the census & housing examples.

Please note that, I am trying to present this as a template, not a ready-to-run code. That is, the user will have to modify and change parts of it before it can be used. And depending of the specificity of the user's model, these change may include only the metadata file, or also other modules to customise data reading, feature processing, feature column creations (if he/she doesn't like the current approach). The value added is the adaptability and standardisation, in my opinion, as well as a quick and simple start for standard data/tasks.

That being said, I kindly recommend the following action items:
**my side**
- Write a detailed walk-through example in the README file to describe how to adapt the template to Census.
- Change the README title to **Cloud MLE - TF Estimator Trainer Package Template, and make sure that I clarify the idea that it is only suitable for the "estimator" APIs in the README file.

**your side**
- Kindly let me know if the example are useful, and if my description make sense.
- Kindly review the template and their related examples, and let me know what can be improved.
- Kindly let me know if you agree with the recommended next steps, and/or if you recommend other ones.

Thanks

**Comment by nkashy1** [comment]
@ksalama : Your recommendations sound good to me.

I like your idea of presenting it as an "estimator on ML Engine" guide. My impression of how people use this repository is that they duplicate the example closest to their use case and start modifying it. We should eventually add your image and text analysis examples to this repo, but it is probably better to create an issue and do it as a separate PR as this one is already quite large.

Am a bit busy with a conference tomorrow, but will look also go through the baby weight custom estimator code at the end of the week and give more detailed feedback on the template. Apologies about the long turn-around.

Thank you.

**Comment by ksalama** [comment]
@nkashy1 : Thank you so much for this.

Just a small note: the baby weight custom estimator might not show all the features covered by the template. For example, process features as part of reading data, creating extended feature columns, using pre-computed stats to normalise numerical columns, instance weighting, etc. are shown in the Census and Housing examples. So you may also like to have a look onto these ones. 

Cheers.

**Comment by ksalama** [comment]
@nkashy1 : Thank you so much for you feedback. The following is a summary to the current changes I added to the PR:

1. **Update Readme.md** - I have added a new section in the Readme file to describe the repo structure. I referred to the original source of the dataset, similar to the [Readme.md ](https://github.com/ksalama/cloudml-samples/blob/master/census/README.md)of the Census example (which is already published). Note that, the only difference is that, in the Census example, the data is uploaded to GCP (which is not different from the data right perspective, as long as you are referencing the original source of the data). 

2. **Re-structure of the template directory** I added a "trainer" sub-directory to the template directory, where I included the template modules. I also added a setup.py and config.yaml files to the template directory. Now the "template" directory acts as the project root, which includes the "trainer" package, and other files.

3. **An implementation of the exponential decay of learning rate in the custom estimator template (model_fn)** - By default, the decay rate is set to 1.0, which doesn't have any effect. The comments on top of the function describe how to configure it to use it.

Let me know if these changes (mainly the first one regarding updating the Readme.md) satisfy your change request.
